### PR TITLE
Add CI configuration with GCC and stdexec

### PR DESCRIPTION
**Pull request by aurianer**
I think it would be a good idea to rename the pika variant from p2300 to stdexec, if @msimberg agrees, I'll update that

- [x] Add build CI with stdexec

### CODE REVIEW

**Comment by msimberg** [comment]
cscs-ci run

File: ci/docker/release-cpu-stdexec.yaml [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+  include:
+  - common.yaml
+
+  view: false
+  concretizer:
+    unify:
+      true
+
+  specs:
+    - dla-future@develop +miniapps +ci-test +ci-check-threads ^intel-mkl threads=openmp ^mpich ^pika+stdexec@git.7a47a4aa411c1ca9adfcb152c28cc3dd7b156b4d=main
```
Comment (msimberg) [inline ci/docker/release-cpu-stdexec.yaml:21]:
```suggestion
    - dla-future@develop +miniapps +ci-test +ci-check-threads ^intel-mkl threads=openmp ^mpich ^pika+stdexec ^stdexec@git.7a47a4aa411c1ca9adfcb152c28cc3dd7b156b4d=main
```

File: include/dlaf/sender/traits.h [hunk 2]
```diff
@@ -9,12 +9,17 @@
...
+#include <type_traits>
+
 #include <pika/async_rw_mutex.hpp>
 #include <pika/execution.hpp>
 #include <pika/future.hpp>
 
 #include <dlaf/types.h>
 
+#include "dlaf/types.h"
+
```
Comment (msimberg) [inline include/dlaf/sender/traits.h:22]:
```suggestion
```

File: include/dlaf/sender/traits.h [hunk 3]
```diff
@@ -27,6 +32,54 @@ struct SenderSingleValueTypeImpl<TypeList<TypeList<T>>> {
...
 
+// Utilities to use unique_t
+template <typename T>
+struct is_false : std::integral_constant<bool, !(bool) T::value> {};
+
+template <typename T>
+inline constexpr bool is_false_v = is_false<T>::value;
+
+template <typename... Ts>
+struct always_false : std::false_type {};
```
Comment (msimberg) [inline include/dlaf/sender/traits.h:43]:
```suggestion
// Utilities to use unique_t
template <typename T>
struct IsFalse : std::integral_constant<bool, !(bool) T::value> {};

template <typename T>
inline constexpr bool IsFalse_v = IsFalse<T>::value;

template <typename... Ts>
struct AlwaysFalse : std::false_type {};
```
etc.

File: include/dlaf/sender/traits.h [hunk 4]
```diff
@@ -60,12 +113,15 @@ struct SenderSingleValueTypeImpl<
   using type = RType;
 };
 
+template <typename... Ts>
+using decayed_typelist = TypeList<std::decay_t<Ts>...>;
+
```
Comment (msimberg) [inline include/dlaf/sender/traits.h:118]:
```suggestion
template <typename... Ts>
using DecayedTypeList = TypeList<std::decay_t<Ts>...>;

```

**Review by msimberg — CHANGES_REQUESTED** [review]
Thanks @aurianer for all the work on this!

I'm 99% sure some tests will fail (or more likely hang). If that's the case, let's make this PR only add the build stage with `+stdexec` so that we can get that tested on master as soon as possible, and separately fix the tests. If everything passes, let's do build and test here!

File: include/dlaf/sender/traits.h [hunk 5]
```diff
@@ -60,12 +111,15 @@ struct SenderSingleValueTypeImpl<
   using type = RType;
 };
 
+template <typename... Ts>
+using DecayedTypeList = TypeList<std::decay_t<Ts>...>;
```
Comment (msimberg) [inline include/dlaf/sender/traits.h:115]: Maybe you can group this together with `TypeList`?

File: include/dlaf/sender/traits.h [hunk 6]
```diff
@@ -27,7 +30,55 @@ struct SenderSingleValueTypeImpl<TypeList<TypeList<T>>> {
   using type = T;
 };
 
-struct empty_env {};
+// Utilities to use unique_t
```
Comment (msimberg) [inline include/dlaf/sender/traits.h:33]: Leave out or update to `UniqueType`? The utilities may be useful on their own at some point so my preference would be to leave it out.

**Review by msimberg — APPROVED** [review]
This looks good to me, thanks @aurianer! Just a couple of minor optional cosmetic comments.

File: include/dlaf/sender/traits.h [hunk 7]
```diff
@@ -19,6 +22,58 @@ namespace dlaf::internal {
...
+struct Unique;
+
+template <template <typename...> class Pack, typename... Ts>
+struct Unique<Pack<Ts...>> : UniqueHelper<Pack<>, Pack<Ts...>> {};
+
+/// Remove duplicate types in the given pack.
+template <typename Pack>
+using UniqueType = typename Unique<Pack>::type;
+
+struct EmptyEnv {};
```
Comment (rasolca) [inline include/dlaf/sender/traits.h:75]: Just for curiosity:
Why are these struct needed for stdexec?

**Review by rasolca — APPROVED** [review]
LGTM

File: include/dlaf/sender/traits.h (hunk 7 above)
Comment (msimberg) [reply include/dlaf/sender/traits.h:75]: It's technically also needed when stdexec is disabled in pika _in the general case_. However, practically we don't need the decaying in the use cases that currently exist in DLA-Future.

The reason is that some stdexec sender adaptors send completion signatures with `T&&` where pika would send `T` and we need to make sure we treat the two the same (as `T`).

**Pull request merged**